app/lib/preload.js

```diff
@@ -9,29 +9,82 @@
  */
 
 const {
-  remote,
+  contextBridge,
   ipcRenderer
 } = require('electron');
 
-const {
-  process,
-  app
-} = remote;
+const allowedEvents = [
+  'errorTracking:turnedOn',
+  'errorTracking:turnedOff',
+  'external:open-url',
+  'dialog:open-files',
+  'dialog:open-file-error',
+  'dialog:save-file',
+  'dialog:show',
+  'file:read',
+  'file:read-stats',
+  'file:write',
+  'zeebe:checkConnection',
+  'zeebe:deploy',
+  'zeebe:run',
+  'config:get',
+  'config:set',
+  'toggle-plugins',
+  'client:ready',
+  'client:error',
+  'app:quit-allowed',
+  'workspace:restore',
+  'workspace:save',
+  'menu:register',
+  'menu:update',
+  'context-menu:open'
+];
 
-const {
-  platform
-} = process;
+const api = {
+  send: (event, ...args) => {
+    if (!allowedEvents.includes(event)) {
+      throw new Error(`Disallowed event: ${event}`);
+    }
+
+    ipcRenderer.send(event, ...args);
+  },
+  on: (event, callback) => {
+    ipcRenderer.on(event, callback);
+
+    return {
+      cancel() {
+        ipcRenderer.off(event, callback);
+      }
+    };
+  },
+  once: (event, callback) => {
+    ipcRenderer.once(event, callback);
+
+    return {
+      cancel() {
+        ipcRenderer.off(event, callback);
+      }
+    };
+  }
+};
+
+let executed = false;
+
+contextBridge.exposeInMainWorld('getAppPreload', function() {
```

Comment:
What is the use case for this one? I.e. how can it be executed more than once?

Reply:
This is to prevent potential attacker from calling `window.getAppPreload`. Within the app logic, there is no reason why it should be allowed to call it multiple times.

Also, cf. https://github.com/camunda/camunda-modeler/blob/v4.7.0/client/src/remote/__tests__/RemoteSpec.js
This is the same logic but moved to `preload.js`.

Reply:
Understood. Did not realize we're actually exposing that function to call. Never mind my previous comment.

Reply:
No problem.